custom-api/store_server.py

```diff
@@ -0,0 +1,296 @@
+#
+# Arguments:
+#   store_id - Required ID of the store supplied in the URL.
+#   request_id - Optional ID used to poll for status on a request.
+#   chunk_index - Optional chunk index that can be used to get an external link to data for a successful request.
+#
+# Return Values:
+#   request_id - Returned for all successfully accepted requests. Used this ID to poll for completion status.
+#   state - Returned on the initial request and status polls. Can be PENDING, RUNNING, SUCCEEDED, FAILED, CANCELED,
+#           or CLOSED.
+#   chunk_count - Returned for successful requests, indicating how many chunks of sales data there are for the store.
+#   link - A pre-signed URL returned when a request_id and chunk_index are supplied as arguments.
+#   
+@app.route("/api/1.0/stores/<string:store_id>/sales", methods=["GET"])
+def list_sales_async(store_id):
+    # If a request ID is not supplied, execute a new list request.
+    request_id = request.args.get('request_id')
+    chunk_index = request.args.get('chunk_index')
+    if request_id is None:
+        # Get the format from the request. It can be ‘CSV’, ‘JSON_ARRAY’, or ‘ARROW_STREAM’
+        format = request.args.get('format')
+        row_limit = request.args.get('row_limit')
+        response = execute_list_sales_request(store_id, format, row_limit)
+        
+    # If there is no chunk index, poll for status.
+    elif chunk_index is None:
+        response = get_list_sales_status(request_id)
+
+    # Otherwise return the pre-signed URL for the specified chunk.
+    else:
+        response = get_list_sales_chunk(request_id, chunk_index)
+        
+    return jsonify(response)
+
+# Invoke the Statement Execution API in asynchronous mode, returning an ID so the client can poll on status.
+def execute_list_sales_request(store_id, format, row_limit):
+    # Use parameters to prevent SQL injection via the store ID string.
+    parameters = [
+        StatementParameterListItem(name='store_id', value=store_id, type="INT")
+    ]
+
+    statement_response = w.statement_execution.execute_statement(
+        statement = sql_statements[SqlStatement.LIST_SALES],
+        format = Format[format],
+        disposition = Disposition.EXTERNAL_LINKS,
+        wait_timeout = "0s",
+        warehouse_id = warehouse_id,
+        parameters = parameters,
+        row_limit = row_limit,
+        byte_limit = 500000000 # 500MB limit because lcp crashes.
+        )
+
+    response = {
+        'request_id': statement_response.statement_id,
+        'state': str(statement_response.status.state.name)
+    }
+
+    return response
+
+# Poll the Statement Execution API for status on a previously generated execution request.
+def get_list_sales_status(request_id):
+    statement_response = w.statement_execution.get_statement(statement_id = request_id)
+    response = {
+        'request_id': statement_response.statement_id,
+        'state': str(statement_response.status.state.name)
```

Comment:
Because we're not returning the result data here (which is available) it costs another round trip to fetch the first chunk.